File: scripts/docs-check.py
#!/usr/bin/env python3
from __future__ import annotations

import json
import re
import sys
from pathlib import Path
from urllib.parse import urlparse


ALLOWED_DOC_TYPES = {
    "overview",
    "architecture",
    "guide",
    "reference",
    "troubleshooting",
    "decision",
    "runbook",
}
REQUIRED_FIELDS = {"doc-schema-version", "title", "summary", "read_when", "doc_type"}
EXCLUDED_DIRS = {"archive", "research"}
MARKDOWN_LINK_RE = re.compile(r"(?<!!)\[[^\]]+\]\(([^)\s]+)(?:\s+\"[^\"]*\")?\)")


def strip_quotes(value: str) -> str:
    value = value.strip()
    if len(value) >= 2 and value[0] == value[-1] and value[0] in {"'", '"'}:
        return value[1:-1]
    return value


def compact_strings(values: list[object]) -> list[str]:
    out: list[str] = []
    for value in values:
        if value is None:
            continue
        text = str(value).strip()
        if text:
            out.append(text)
    return out


def parse_inline_list(value: str) -> list[str]:
    try:
        parsed = json.loads(value.replace("'", '"'))
    except json.JSONDecodeError:
        return []
    return compact_strings(parsed) if isinstance(parsed, list) else []


def parse_frontmatter(path: Path) -> tuple[dict[str, object], str | None]:
    raw = path.read_text(encoding="utf-8")
    if not raw.startswith("---\n") and not raw.startswith("---\r\n"):
        return {}, "missing front matter"

    lines = raw.splitlines()
    end_index = None
    for index, line in enumerate(lines[1:], start=1):
        if line.strip() in {"---", "..."}:
            end_index = index
            break

    if end_index is None:
        return {}, "unterminated front matter"

    data: dict[str, object] = {}
    collecting: str | None = None
    for raw_line in lines[1:end_index]:
        line = raw_line.strip()
        if not line:
            continue

        if collecting and line.startswith("- "):
            current = data.setdefault(collecting, [])
            if isinstance(current, list):
                current.append(strip_quotes(line[2:].strip()))
            continue

        collecting = None
        if ":" not in line:
            continue

        key, value = line.split(":", 1)
        key = key.strip()
        value = value.strip()
        if not value:
            data[key] = []
            collecting = key
        elif value.startswith("[") and value.endswith("]"):
            data[key] = parse_inline_list(value)
        else:
            data[key] = strip_quotes(value)

    return data, None


def walk_docs(docs_dir: Path) -> list[Path]:
    files: list[Path] = []
    for path in docs_dir.rglob("*.md"):
        if any(part.startswith(".") or part in EXCLUDED_DIRS for part in path.relative_to(docs_dir).parts):
            continue
        files.append(path)
    return sorted(files, key=lambda item: item.relative_to(docs_dir).as_posix())


def route_for(relative_path: str) -> set[str]:
    stem = re.sub(r"\.md$", "", relative_path)
    routes = {normalize_route(stem)}
    if stem == "index":
        routes.add("/")
    if stem.endswith("/index"):
        routes.add(normalize_route(stem[: -len("/index")]))
    return routes


def normalize_route(route: str) -> str:
    route = route.split("#", 1)[0].split("?", 1)[0].strip("/")
    return f"/{route}" if route else "/"


def collect_nav_pages(value: object) -> list[str]:
    pages: list[str] = []
    if isinstance(value, list):
        for item in value:
            pages.extend(collect_nav_pages(item))
    elif isinstance(value, dict):
        for key, item in value.items():
            if key == "pages" and isinstance(item, list):
                for page in item:
                    if isinstance(page, str):
                        pages.append(page)
                    else:
                        pages.extend(collect_nav_pages(page))
            else:
                pages.extend(collect_nav_pages(item))
    return pages


def check_frontmatter(path: Path, docs_dir: Path) -> list[str]:
    rel = path.relative_to(docs_dir).as_posix()
    frontmatter, parse_error = parse_frontmatter(path)
    errors: list[str] = []
    if parse_error:
        return [f"{rel}: {parse_error}"]

    missing = sorted(REQUIRED_FIELDS - set(frontmatter))
    for field in missing:
        errors.append(f"{rel}: missing frontmatter field `{field}`")

    schema_version = str(frontmatter.get("doc-schema-version", "")).strip()
    if schema_version != "1":
        errors.append(f"{rel}: `doc-schema-version` must be 1")

    for field in ["title", "summary", "doc_type"]:
        value = str(frontmatter.get(field, "")).strip()
        if not value:
            errors.append(f"{rel}: `{field}` must be non-empty")

    doc_type = str(frontmatter.get("doc_type", "")).strip()
    if doc_type and doc_type not in ALLOWED_DOC_TYPES:
        allowed = ", ".join(sorted(ALLOWED_DOC_TYPES))
        errors.append(f"{rel}: `doc_type` must be one of: {allowed}")

    read_when = frontmatter.get("read_when", [])
    if not isinstance(read_when, list) or not compact_strings(read_when):
        errors.append(f"{rel}: `read_when` must contain at least one hint")

    return errors


def check_code_fences(path: Path, docs_dir: Path) -> list[str]:
    rel = path.relative_to(docs_dir).as_posix()
    fence_count = 0
    for line in path.read_text(encoding="utf-8").splitlines():
        if line.lstrip().startswith("```"):
            fence_count += 1
    return [f"{rel}: unbalanced fenced code block"] if fence_count % 2 else []


def check_docs_json(docs_dir: Path, markdown_files: list[Path]) -> list[str]:
    config_path = docs_dir / "docs.json"
    if not config_path.exists():
        return ["docs/docs.json: missing docs navigation file"]

    try:
        config = json.loads(config_path.read_text(encoding="utf-8"))
    except json.JSONDecodeError as error:
        return [f"docs/docs.json: invalid JSON at line {error.lineno}: {error.msg}"]

    errors: list[str] = []
    page_paths = {path.relative_to(docs_dir).as_posix() for path in markdown_files}
    page_keys = {re.sub(r"\.md$", "", path) for path in page_paths}
    for page in collect_nav_pages(config.get("navigation", [])):
        key = page.strip().strip("/")
        if key not in page_keys:
            errors.append(f"docs/docs.json: navigation page `{page}` has no matching docs page")

    return errors


def is_external(target: str) -> bool:
    parsed = urlparse(target)
    return bool(parsed.scheme) or target.startswith(("mailto:", "tel:"))


def check_links(path: Path, docs_dir: Path, routes: set[str]) -> list[str]:
    rel = path.relative_to(docs_dir).as_posix()
    text = path.read_text(encoding="utf-8")
    errors: list[str] = []

    for match in MARKDOWN_LINK_RE.finditer(text):
        target = match.group(1).strip()
        if not target or target.startswith("#") or is_external(target):
            continue

        target_path = target.split("#", 1)[0].split("?", 1)[0]
        if target_path.startswith("/"):
            route = normalize_route(target_path)
            if route not in routes:
                errors.append(f"{rel}: broken docs route `{target}`")
            continue

        resolved = (path.parent / target_path).resolve()
        if not str(resolved).startswith(str(docs_dir.resolve())):
            continue
        if target_path.endswith(".md") and not resolved.exists():
            errors.append(f"{rel}: broken relative link `{target}`")

    return errors


def main() -> int:
    docs_dir = Path("docs")
    if not docs_dir.exists() or not docs_dir.is_dir():
        print("docs:check: missing docs directory. Run from repo root.", file=sys.stderr)
        return 1

    markdown_files = walk_docs(docs_dir)
    routes: set[str] = set()
    for path in markdown_files:
        routes.update(route_for(path.relative_to(docs_dir).as_posix()))

    errors: list[str] = []
    errors.extend(check_docs_json(docs_dir, markdown_files))
    for path in markdown_files:
        errors.extend(check_frontmatter(path, docs_dir))
        errors.extend(check_code_fences(path, docs_dir))
        errors.extend(check_links(path, docs_dir, routes))

    if errors:
        for error in errors:
            print(f"docs:check: {error}", file=sys.stderr)
        return 1

    print(f"docs:check: ok ({len(markdown_files)} docs page(s))")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/docs-check.mjs
#!/usr/bin/env node
// Validate the docs tree: frontmatter schema, balanced code fences, navigation entries that
// resolve to a real page, and internal links that resolve to a real route or file.
//
// Runs in `docs:check` and again inside `docs:build`, so a broken link fails the docs build
// rather than shipping a 404.
//
// With `--dist`, checks the built output instead: every URL in the generated llms.txt files
// must name a page VitePress actually emitted. That is a separate pass because it can only run
// *after* `vitepress build`, and the source checks above only need the markdown tree.

import { existsSync, readFileSync } from 'node:fs'
import path from 'node:path'

import { ALLOWED_DOC_TYPES, compactStrings, parseFrontmatter, requireDocsDir, walkDocs } from './docs-frontmatter.mjs'

const REQUIRED_FIELDS = ['doc-schema-version', 'doc_type', 'read_when', 'summary', 'title']
// Markdown inline links, excluding images (`!` prefix) and tolerating a trailing "title".
const MARKDOWN_LINK_RE = /(?<!!)\[[^\]]+\]\(([^)\s]+)(?:\s+"[^"]*")?\)/g

/** Normalize a site route: strip query/hash and surrounding slashes, keeping a leading `/`. */
function normalizeRoute(route) {
	const bare = (route.split('#', 1)[0]?.split('?', 1)[0] ?? '').replace(/^\/+|\/+$/g, '')
	return bare ? `/${bare}` : '/'
}

/** Every route a docs page is reachable at (a page, plus `/` for index pages). */
function routesFor(relativePath) {
	const stem = relativePath.replace(/\.md$/, '')
	const routes = new Set([normalizeRoute(stem)])
	if (stem === 'index') routes.add('/')
	if (stem.endsWith('/index')) routes.add(normalizeRoute(stem.slice(0, -'/index'.length)))
	return routes
}

/** Every `pages:` entry anywhere in a docs.json navigation tree. */
function collectNavPages(value) {
	const pages = []
	if (Array.isArray(value)) {
		for (const item of value) pages.push(...collectNavPages(item))
	} else if (value && typeof value === 'object') {
		for (const [key, item] of Object.entries(value)) {
			if (key === 'pages' && Array.isArray(item)) {
				for (const page of item) {
					if (typeof page === 'string') pages.push(page)
					else pages.push(...collectNavPages(page))
				}
			} else {
				pages.push(...collectNavPages(item))
			}
		}
	}
	return pages
}

/** A link target that leaves the docs site entirely. */
function isExternal(target) {
	return /^[a-z][a-z0-9+.-]*:/i.test(target) || target.startsWith('//')
}

function checkFrontmatter(docsDir, rel) {
	const { data, error } = parseFrontmatter(path.join(docsDir, rel))
	if (error) return [`${rel}: ${error}`]

	const errors = []
	for (const field of REQUIRED_FIELDS) {
		if (!Object.hasOwn(data, field)) errors.push(`${rel}: missing frontmatter field \`${field}\``)
	}
	if (String(data['doc-schema-version'] ?? '').trim() !== '1') errors.push(`${rel}: \`doc-schema-version\` must be 1`)
	for (const field of ['title', 'summary', 'doc_type']) {
		if (!String(data[field] ?? '').trim()) errors.push(`${rel}: \`${field}\` must be non-empty`)
	}
	const docType = String(data.doc_type ?? '').trim()
	if (docType && !ALLOWED_DOC_TYPES.includes(docType)) {
		errors.push(`${rel}: \`doc_type\` must be one of: ${ALLOWED_DOC_TYPES.join(', ')}`)
	}
	if (compactStrings(data.read_when).length === 0) errors.push(`${rel}: \`read_when\` must contain at least one hint`)
	return errors
}

function checkCodeFences(docsDir, rel) {
	const text = readFileSync(path.join(docsDir, rel), 'utf8')
	const fences = text.split(/\r?\n/).filter((line) => line.trimStart().startsWith('```')).length
	return fences % 2 ? [`${rel}: unbalanced fenced code block`] : []
}

function checkDocsJson(docsDir, relPaths) {
	const configPath = path.join(docsDir, 'docs.json')
	if (!existsSync(configPath)) return ['docs/docs.json: missing docs navigation file']

	let config
	try {
		config = JSON.parse(readFileSync(configPath, 'utf8'))
	} catch (error) {
		return [`docs/docs.json: invalid JSON: ${error instanceof Error ? error.message : String(error)}`]
	}

	const pageKeys = new Set(relPaths.map((rel) => rel.replace(/\.md$/, '')))
	const errors = []
	for (const page of collectNavPages(config.navigation ?? [])) {
		const key = page.trim().replace(/^\/+|\/+$/g, '')
		if (!pageKeys.has(key)) errors.push(`docs/docs.json: navigation page \`${page}\` has no matching docs page`)
	}
	return errors
}

function checkLinks(docsDir, rel, routes) {
	const filePath = path.join(docsDir, rel)
	const text = readFileSync(filePath, 'utf8')
	const errors = []
	const docsRoot = path.resolve(docsDir)

	for (const match of text.matchAll(MARKDOWN_LINK_RE)) {
		const target = (match[1] ?? '').trim()
		if (!target || target.startsWith('#') || isExternal(target)) continue

		const targetPath = target.split('#', 1)[0]?.split('?', 1)[0] ?? ''
		if (targetPath.startsWith('/')) {
			if (!routes.has(normalizeRoute(targetPath))) errors.push(`${rel}: broken docs route \`${target}\``)
			continue
		}
		const resolved = path.resolve(path.dirname(filePath), targetPath)
		if (!resolved.startsWith(docsRoot)) continue // points outside the docs tree; not ours to check
		if (targetPath.endsWith('.md') && !existsSync(resolved)) errors.push(`${rel}: broken relative link \`${target}\``)
	}
	return errors
}

/**
 * The URL prefix the published site actually answers on, derived rather than restated: the
 * GitHub Pages host comes from `repository`, the path from the VitePress `base`. Writing the
 * value out a second time is what let `llms.txt` drift to a host that never existed.
 */
function canonicalBase(docsDir) {
	if (process.env.DOCS_BASE_URL) return { base: process.env.DOCS_BASE_URL.replace(/\/?$/, '/'), errors: [] }

	const pkg = JSON.parse(readFileSync(path.join(docsDir, '..', 'package.json'), 'utf8'))
	const slug = String(pkg.repository?.url ?? pkg.homepage ?? '').match(/github\.com\/([^/]+)\/([^/.#]+)/)
	if (!slug) return { base: null, errors: ['package.json: cannot derive the GitHub Pages host from `repository`'] }

	const config = readFileSync(path.join(docsDir, '.vitepress', 'config.mts'), 'utf8')
	const configured = config.match(/^\s*base:.*?'([^']+)'/m)
	if (!configured) return { base: null, errors: ['docs/.vitepress/config.mts: cannot read the `base` option'] }

	// An unreadable `base` collapses to `/`, which the slug check below then rejects loudly.
	const base = (process.env.VITEPRESS_BASE ?? configured[1] ?? '').replace(/\/?$/, '/')
	const errors = []
	// A project Pages site is served under /<repo>/, so a `base` that disagrees means the whole
	// site 404s no matter how well-formed the routes underneath it are.
	if (!process.env.VITEPRESS_BASE && base !== `/${slug[2]}/`) {
		errors.push(`docs/.vitepress/config.mts: \`base\` is \`${base}\`, but Pages serves this repo at \`/${slug[2]}/\``)
	}
	return { base: `https://${slug[1]}.github.io${base}`, errors }
}

// How each generated file writes the URLs it advertises. These have to stay per-file rather than
// being tried against both: llms-full.txt embeds whole page bodies, so a link pattern applied to
// it would also match the ordinary markdown links *inside* those bodies.
const ADVERTISED_URL_RE = {
	'llms-full.txt': /^URL: (\S+)$/gm,
	'llms.txt': /^- \[[^\]]+\]\((\S+?)\)/gm,
}

/** The URLs a generated llms file advertises. */
function advertisedUrls(text, pattern) {
	const urls = [...text.matchAll(pattern)].map((match) => match[1] ?? '')
	return new Set(urls.map((url) => url.split('#', 1)[0]?.split('?', 1)[0] ?? ''))
}

/**
 * Every advertised URL must map to a file in the build. VitePress runs with `cleanUrls`, so a
 * leaf page is `x.html` and only a directory index is `x/index.html` — the mapping below is the
 * one the server applies, so a route the generator invents cannot pass by looking plausible.
 */
function checkGeneratedUrls(docsDir) {
	const dist = path.join(docsDir, '.vitepress', 'dist')
	if (!existsSync(dist)) return ['docs/.vitepress/dist: missing; run `vitepress build docs` before `--dist`']

	const { base, errors } = canonicalBase(docsDir)
	if (!base) return errors

	for (const [name, pattern] of Object.entries(ADVERTISED_URL_RE)) {
		const file = path.join(dist, name)
		if (!existsSync(file)) {
			errors.push(`${name}: not present in the build output`)
			continue
		}
		const urls = advertisedUrls(readFileSync(file, 'utf8'), pattern)
		if (urls.size === 0) {
			errors.push(`${name}: advertises no URLs, so nothing was verified`)
			continue
		}
		for (const url of [...urls].sort()) {
			if (!url.startsWith(base)) {
				errors.push(`${name}: \`${url}\` is not under the published base \`${base}\``)
				continue
			}
			const rest = url.slice(base.length)
			const page = rest === '' ? 'index.html' : rest.endsWith('/') ? `${rest}index.html` : `${rest}.html`
			if (!existsSync(path.join(dist, page))) errors.push(`${name}: \`${url}\` has no page in the build (${page})`)
		}
	}
	return errors
}

const docsDir = requireDocsDir('docs:check')

if (process.argv.includes('--dist')) {
	const errors = checkGeneratedUrls(docsDir)
	if (errors.length > 0) {
		for (const error of errors) console.error(`docs:check: ${error}`)
		process.exit(1)
	}
	console.log('docs:check: ok (generated llms URLs all resolve to a built page)')
	process.exit(0)
}

const relPaths = walkDocs(docsDir)

const routes = new Set()
for (const rel of relPaths) for (const route of routesFor(rel)) routes.add(route)

const errors = [...checkDocsJson(docsDir, relPaths)]
for (const rel of relPaths) {
	errors.push(...checkFrontmatter(docsDir, rel), ...checkCodeFences(docsDir, rel), ...checkLinks(docsDir, rel, routes))
}

if (errors.length > 0) {
	for (const error of errors) console.error(`docs:check: ${error}`)
	process.exit(1)
}
console.log(`docs:check: ok (${relPaths.length} docs page(s))`)
